Use `plist-get` and `plist-put` with `#'equal` from Emacs/Compat 29

meliache opened the pull request:
From Emacs 29 (currently in pretest), `plist-get` and `plist-put` take an optional comparison predicate which we set to `#'equal`. Previously, `lax-plist-get`/`-put` was used for that, but those functions are now marked as obsolete. They still work as wrapper, but give obsoletion warnings. E.g. the new `lax-plist-get` implementation is

``` elisp
(defun lax-plist-get (plist prop)
  "Extract a value from a property list, comparing with `equal'."
  (declare (pure t) (side-effect-free t) (obsolete plist-get "29.1"))
  (plist-get plist prop #'equal))
```

To keep forward-compatibility with older Emacs releases, I added a dependency on [compat](https://elpa.gnu.org/packages/compat.html), which implements the Emacs-29 version of the plist functions and should take care of that, but I didn't explicitly test older releases yet. `compat`  is getting quite whitespread, so I think it's reasonable to add that dependency.

meliache committed:
Use `plist-get` and `plist-put` with `#'equal` from Emacs/Compat 29

From Emacs 29 (currently in pretest), `plist-get` and `plist-put` take an
optional comparison predicate which we set to `#'equal`. Previously,
`lax-plist-get`/`-put` was used for that, but those functions are now marked as
obsolete.

To keep compatibility with older Emacs releases, I added a dependency on
Compat (https://elpa.gnu.org/packages/compat.html).

jerrypnz commented:
Thanks @meliache for yet another PR. The change looks good to me but I'll test it out on my Emacs 28 and then merge it.

jerrypnz commented:
Hmm this seems to break on emacs 28. It complains about the additional argument. Looking at the [compat manual](https://elpa.gnu.org/packages/doc/compat.html) (search `plist-get`), I wonder if you're supposed to call compat functions this way:

```elisp
(compat-call plist-get plist prop #'equal)
```

meliache commented:
Your solution with `compat-call` is correct. I just looked at the docstring of `compat-call`:
```
Call compatibility function or macro FUN with ARGS.

A good example function is plist-get which was extended with an
additional predicate argument in Emacs 29.1.
[...]
```
:facepalm: 
I'll implement that and then install Emacs 28 and try it myself. Sorry for not doing that in the first place. Initially I had just looked at some example packages that use compat like  [jinx.el](https://github.com/minad/jinx/blob/main/jinx.el) and they seemed to do nothing else then just adding the requirements, there I didn't find any `compat-call` functions, probably because it's not necessary when compat is just used for accessing a function that was introduced in newer Emacs versions but didn't exist previously.

The downside is that this PR makes the code a bit more verbose and less readable...

meliache committed:
Use compat-call to use plist-funcs from compat.el

compat.el newer overrides existing functions. Therefore, for using new versions
of existing functions like `plist-get` and `plist-put` with custom
testfunctions, `compat-call` needs to be used.

meliache commented:
By the way, with some minor effort I think one could get rid of the `s` and `dash` dependencies, as many of their functions have since then been built-in into recent emacs releases and on older versions they are accessible via `compat`. E.g.
| libary function | builtin/compat |
|-|-|
|`->>` |`thread-last`|
| `s-pad-right` | `string-pad`|
| `s-replace` | `string-replace`'
| `-map` | `seq-map`|
| `-group-by` | `seq-group-by`|
| ... | ... |

But to be honest I don't care that much, since I have `s` and `dash` installed anyway due to other packages requiring those and they are not that large and you might want to keep them, e.g. due to `->>` being shorter and more clojure-like than `thread-last`.

jerrypnz commented:
Sorry for the late response. It seems that there are some conflicts with master branch. If you can take a look and resolve it, I'm happy to merge the change. As for dash.el, I'm a Clojure developer so it's probably not surprising to you why I chose it. I have a feeling that it is in people's list of packages anyway so doesn't hurt to use it.

meliache committed:
Merge remote-tracking branch 'origin/master' into emacs-29-plist-compat

meliache commented:
Upps that is the downside to slitting changes into multiple PR's. Merged master and resolved the conflict.

jerrypnz commented on major-mode-hydra.el:
```diff
@@ -75,7 +75,7 @@ Overwrite existing hydra if OVERWRITE-P is t, otherwise add new heads to it."
                   (funcall major-mode-hydra-title-generator mode)))
          (body (-> body
-                   (lax-plist-put :hint nil)
+                   (compat-call plist-put :hint nil #'equal)
```
Sorry for the super late response. Work has been busy and I kinda forgot about this PR. This bit breaks because of the `->` macro. It gets expanded to `(compat-call body plist-put :hint nil #'equal)` which is wrong. I'll merge your PR and fix it on `master`.

jerrypnz commented:
@meliache  Seems to work fine on both emacs 28 and 29 after I fixed that one issue. Thanks for the contribution and sorry for my super late response.
